{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f6adce82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# autoreload modules; useful for testing\n",
    "# %load_ext autoreload\n",
    "# %autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a04dfc56-7eec-49fb-a593-571d78bcbb13",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import copy\n",
    "import pandas as pd\n",
    "from toolz.dicttoolz import assoc_in, merge, dissoc\n",
    "from dotenv import dotenv_values\n",
    "from pymongo import MongoClient\n",
    "from pymongo.database import Database as MongoDatabase\n",
    "from nmdc_runtime.api.core.metadata import load_changesheet, update_mongo_db, mongo_update_command_for, copy_docs_in_update_cmd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30be0a13",
   "metadata": {},
   "source": [
    "### load mongodb via env info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c4208227",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mongodb://localhost:27018'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = dotenv_values(\"../../.env.localhost\")\n",
    "config[\"MONGO_HOST\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae247da1-ec7e-4fe2-b713-d0a4e97e6a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "client = MongoClient(host=config[\"MONGO_HOST\"], username=config[\"MONGO_USERNAME\"], password=config[\"MONGO_PASSWORD\"])\n",
    "mongodb = client[\"nmdc_etl_staging\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c97d66c0-1aff-4fe3-abd8-5161787ea682",
   "metadata": {},
   "source": [
    "### create temp database and temp_set collection for testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "91c95437-00fe-40a2-94a8-2210a27a1741",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"temp_db\" in client.list_database_names():\n",
    "    client.drop_database(\"temp_db\")\n",
    "temp_db = client[\"temp_db\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91901ba-be06-4ad7-a7ef-ea614167a8d9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### helper functons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f16aec90-b722-487d-982f-3780d8d052ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# wraps the mongo_update_command_for and update_mongo_db into\n",
    "# a single function to process the change sheet\n",
    "def process_changesheet(changeDf, mdb: MongoDatabase, temp_db: MongoDatabase):\n",
    "    update_cmd = mongo_update_command_for(changeDf)\n",
    "    copy_docs_in_update_cmd(update_cmd, mdb, temp_db)\n",
    "    return update_mongo_db(temp_db, update_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7973a575-dca9-4ed2-9bc8-1d169f6cc97d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# puts the change sheet results in dataframe\n",
    "def print_results(results):\n",
    "    for i, result in enumerate(results):\n",
    "        print(f\"\\n============== {result['id']} ==============\")\n",
    "        print(\"------------------ BEFORE ------------------\")\n",
    "        print(json.dumps(result[\"doc_before\"], indent=2))\n",
    "        print(\"------------------ AFTER ------------------\")\n",
    "        print(json.dumps(result[\"doc_after\"], indent=2))\n",
    "        print(\"------------------ ERRORS ------------------\")\n",
    "        print(\"\\n\".join(result[\"validation_errors\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8e21247-e412-441d-b608-3704b91e3a58",
   "metadata": {},
   "source": [
    "### set dataframe display options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1c15519c-81c2-4672-820b-f08bf6884a6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_columns\", None)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb81d97-df2a-411f-8f15-fa4b5fcbdbff",
   "metadata": {},
   "source": [
    "### process change sheet 1\n",
    "Simple test with two studies. The doi is changed using \".\" path separator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33b2995a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>action</th>\n",
       "      <th>attribute</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gold:Gs0114663</td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>UPDATED NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem</td>\n",
       "      <td>SOIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gold:Gs0103573</td>\n",
       "      <td>update</td>\n",
       "      <td>doi.has_raw_value</td>\n",
       "      <td>10.9999/8888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>UPDATED NAME 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  action          attribute           value\n",
       "0  gold:Gs0114663  update               name  UPDATED NAME 1\n",
       "1                  update          ecosystem            SOIL\n",
       "2  gold:Gs0103573  update  doi.has_raw_value    10.9999/8888\n",
       "3                  update               name  UPDATED NAME 2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/changesheet-with-separator1.tsv\", sep=\"\\t\", dtype=\"string\").fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7ed229bc-4d72-4101-af9d-d001b1af6329",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sheetDf = load_changesheet(\"data/changesheet-with-separator1.tsv\", mongodb)\n",
    "# sheetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd26f39c-cb7d-4b17-ab2d-757f26017738",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============== gold:Gs0103573 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"Populus root and rhizosphere microbial communities from Tennessee, USA\",\n",
      "  \"description\": \"Defining the functional diversity of the Populus root microbiome\",\n",
      "  \"ecosystem\": \"Host-associated\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"Unclassified\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Dale Pelletier\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"UPDATED NAME 2\",\n",
      "  \"description\": \"Defining the functional diversity of the Populus root microbiome\",\n",
      "  \"ecosystem\": \"Host-associated\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"Unclassified\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Dale Pelletier\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\",\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.9999/8888\"\n",
      "  }\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n",
      "\n",
      "============== gold:Gs0114663 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"Riverbed sediment microbial communities from the Columbia River, Washington, USA\",\n",
      "  \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone (hyporheic zone).\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"Freshwater\",\n",
      "  \"ecosystem_subtype\": \"River\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"James Stegen\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487765\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"UPDATED NAME 1\",\n",
      "  \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone (hyporheic zone).\",\n",
      "  \"ecosystem\": \"SOIL\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"Freshwater\",\n",
      "  \"ecosystem_subtype\": \"River\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"James Stegen\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487765\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n"
     ]
    }
   ],
   "source": [
    "print_results(process_changesheet(sheetDf, mongodb, temp_db))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3757667",
   "metadata": {
    "tags": []
   },
   "source": [
    "### process change sheet 2\n",
    "Test with two studies, but the second study uses grouping variables/symbols to update object values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c0ba01ce-e3ea-42a6-afd9-27d0b808b96d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>action</th>\n",
       "      <th>attribute</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gold:Gs0103573</td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>NEW NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ecosystem</td>\n",
       "      <td>SOIL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gold:Gs0114663</td>\n",
       "      <td>update</td>\n",
       "      <td>doi</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>v1</td>\n",
       "      <td></td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>10.9999/8888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>NEW NAME 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>principal_investigator</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>v2</td>\n",
       "      <td></td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>NEW RAW VALUE 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  action               attribute            value\n",
       "0  gold:Gs0103573  update                    name       NEW NAME 1\n",
       "1                                       ecosystem             SOIL\n",
       "2  gold:Gs0114663  update                     doi               v1\n",
       "3              v1                   has_raw_value     10.9999/8888\n",
       "4                  update                    name       NEW NAME 2\n",
       "5                  update  principal_investigator               v2\n",
       "6              v2                   has_raw_value  NEW RAW VALUE 2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/changesheet-without-separator1.tsv\", sep=\"\\t\", dtype=\"string\").fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6e5dae80-5799-4315-9a40-722d134d551a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sheetDf = load_changesheet(\"data/changesheet-without-separator1.tsv\", mongodb)\n",
    "# sheetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a8fb5844-ee52-4404-a98b-5b15259386bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============== gold:Gs0103573 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"Populus root and rhizosphere microbial communities from Tennessee, USA\",\n",
      "  \"description\": \"Defining the functional diversity of the Populus root microbiome\",\n",
      "  \"ecosystem\": \"Host-associated\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"Unclassified\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Dale Pelletier\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"NEW NAME 1\",\n",
      "  \"description\": \"Defining the functional diversity of the Populus root microbiome\",\n",
      "  \"ecosystem\": \"SOIL\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"Unclassified\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Dale Pelletier\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n",
      "\n",
      "============== gold:Gs0114663 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"Riverbed sediment microbial communities from the Columbia River, Washington, USA\",\n",
      "  \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone (hyporheic zone).\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"Freshwater\",\n",
      "  \"ecosystem_subtype\": \"River\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"James Stegen\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487765\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"NEW NAME 2\",\n",
      "  \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone (hyporheic zone).\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"Freshwater\",\n",
      "  \"ecosystem_subtype\": \"River\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"NEW RAW VALUE 2\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.9999/8888\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n"
     ]
    }
   ],
   "source": [
    "print_results(process_changesheet(sheetDf, mongodb, temp_db))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee3516ff-77a3-4758-b5d1-eb81dd092002",
   "metadata": {},
   "source": [
    "### process change sheet 3\n",
    "Test with two studies. Both studies are updated using grouping variables/symbols.  \n",
    "Note the reuse of variables/symbols in each id group. A variable/symbol is local the id group it appears in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0e7de6e2-edda-418d-9ddb-9f125809d464",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>action</th>\n",
       "      <th>attribute</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gold:Gs0114663</td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>NEW STUDY NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem</td>\n",
       "      <td>NEW ECOSYSTEM 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem_type</td>\n",
       "      <td>NEW ECOSYSTEM_TYPE 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem_subtype</td>\n",
       "      <td>NEW ECOSYSTEM_SUBTYPE 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>doi</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>v1</td>\n",
       "      <td>update</td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>NEW DOI 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>principal_investigator</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>v2</td>\n",
       "      <td></td>\n",
       "      <td>name</td>\n",
       "      <td>NEW PI NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>v2</td>\n",
       "      <td></td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>NEW RAW NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>description</td>\n",
       "      <td>NEW DESCRIPTION 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>websites</td>\n",
       "      <td>HTTP://TEST1.EXAMPLE.COM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>websites</td>\n",
       "      <td>HTTP://TEST2.EXAMPLE.COM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>has_credit_associations</td>\n",
       "      <td>v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>v3</td>\n",
       "      <td></td>\n",
       "      <td>applied_role</td>\n",
       "      <td>NEW CO-INVESTIGATOR 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>v3</td>\n",
       "      <td></td>\n",
       "      <td>applies_to_person.name</td>\n",
       "      <td>NEW CO-INVESTIGATOR NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>has_credit_associations</td>\n",
       "      <td>v4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>v4</td>\n",
       "      <td></td>\n",
       "      <td>applied_role</td>\n",
       "      <td>NEW CURATOR 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>v4</td>\n",
       "      <td></td>\n",
       "      <td>applies_to_person.name</td>\n",
       "      <td>NEW CURATOR NAME 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gold:Gs0103573</td>\n",
       "      <td>update</td>\n",
       "      <td>name</td>\n",
       "      <td>NEW STUDY NAME 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem</td>\n",
       "      <td>NEW ECOSYSTEM 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem_type</td>\n",
       "      <td>NEW ECOSYSTEM_TYPE 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>ecosystem_subtype</td>\n",
       "      <td>NEW ECOSYSTEM_SUBTYPE 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>doi</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>v1</td>\n",
       "      <td>update</td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>NEW DOI 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>principal_investigator</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>v2</td>\n",
       "      <td></td>\n",
       "      <td>name</td>\n",
       "      <td>NEW PI NAME 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>v2</td>\n",
       "      <td></td>\n",
       "      <td>has_raw_value</td>\n",
       "      <td>NEW RAW NAME 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>description</td>\n",
       "      <td>NEW DESCRIPTION 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>websites</td>\n",
       "      <td>HTTP://TEST3.EXAMPLE.COM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>websites</td>\n",
       "      <td>HTTP://TEST4.EXAMPLE.COM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>has_credit_associations</td>\n",
       "      <td>v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>v3</td>\n",
       "      <td></td>\n",
       "      <td>applied_role</td>\n",
       "      <td>NEW CO-INVESTIGATOR 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>v3</td>\n",
       "      <td></td>\n",
       "      <td>applies_to_person.name</td>\n",
       "      <td>NEW CO-INVESTIGATOR NAME 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td></td>\n",
       "      <td>update</td>\n",
       "      <td>has_credit_associations</td>\n",
       "      <td>v4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>v4</td>\n",
       "      <td></td>\n",
       "      <td>applied_role</td>\n",
       "      <td>NEW CURATOR 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>v4</td>\n",
       "      <td></td>\n",
       "      <td>applies_to_person.name</td>\n",
       "      <td>NEW CURATOR NAME 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id  action                attribute                       value\n",
       "0   gold:Gs0114663  update                     name            NEW STUDY NAME 1\n",
       "1                   update                ecosystem             NEW ECOSYSTEM 1\n",
       "2                   update           ecosystem_type        NEW ECOSYSTEM_TYPE 1\n",
       "3                   update        ecosystem_subtype     NEW ECOSYSTEM_SUBTYPE 1\n",
       "4                   update                      doi                          v1\n",
       "5               v1  update            has_raw_value                   NEW DOI 1\n",
       "6                   update   principal_investigator                          v2\n",
       "7               v2                             name               NEW PI NAME 1\n",
       "8               v2                    has_raw_value              NEW RAW NAME 1\n",
       "9                   update              description           NEW DESCRIPTION 1\n",
       "10                  update                 websites    HTTP://TEST1.EXAMPLE.COM\n",
       "11                  update                 websites    HTTP://TEST2.EXAMPLE.COM\n",
       "12                  update  has_credit_associations                          v3\n",
       "13              v3                     applied_role       NEW CO-INVESTIGATOR 1\n",
       "14              v3           applies_to_person.name  NEW CO-INVESTIGATOR NAME 1\n",
       "15                  update  has_credit_associations                          v4\n",
       "16              v4                     applied_role               NEW CURATOR 1\n",
       "17              v4           applies_to_person.name          NEW CURATOR NAME 1\n",
       "18  gold:Gs0103573  update                     name            NEW STUDY NAME 2\n",
       "19                  update                ecosystem             NEW ECOSYSTEM 2\n",
       "20                  update           ecosystem_type        NEW ECOSYSTEM_TYPE 2\n",
       "21                  update        ecosystem_subtype     NEW ECOSYSTEM_SUBTYPE 2\n",
       "22                  update                      doi                          v1\n",
       "23              v1  update            has_raw_value                   NEW DOI 2\n",
       "24                  update   principal_investigator                          v2\n",
       "25              v2                             name               NEW PI NAME 2\n",
       "26              v2                    has_raw_value              NEW RAW NAME 2\n",
       "27                  update              description           NEW DESCRIPTION 2\n",
       "28                  update                 websites    HTTP://TEST3.EXAMPLE.COM\n",
       "29                  update                 websites    HTTP://TEST4.EXAMPLE.COM\n",
       "30                  update  has_credit_associations                          v3\n",
       "31              v3                     applied_role       NEW CO-INVESTIGATOR 2\n",
       "32              v3           applies_to_person.name  NEW CO-INVESTIGATOR NAME 2\n",
       "33                  update  has_credit_associations                          v4\n",
       "34              v4                     applied_role               NEW CURATOR 2\n",
       "35              v4           applies_to_person.name          NEW CURATOR NAME 2"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/changesheet-without-separator3.tsv\", sep=\"\\t\", dtype=\"string\").fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6b71bd4d-45f1-4b11-890f-65458a85fb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "sheetDf = load_changesheet(\"data/changesheet-without-separator3.tsv\", mongodb)\n",
    "# sheetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5293569c-9590-4ce7-8fe8-5226b37a5679",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============== gold:Gs0103573 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"Populus root and rhizosphere microbial communities from Tennessee, USA\",\n",
      "  \"description\": \"Defining the functional diversity of the Populus root microbiome\",\n",
      "  \"ecosystem\": \"Host-associated\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"Unclassified\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Dale Pelletier\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0103573\",\n",
      "  \"name\": \"NEW STUDY NAME 2\",\n",
      "  \"description\": \"NEW DESCRIPTION 2\",\n",
      "  \"ecosystem\": \"NEW ECOSYSTEM 2\",\n",
      "  \"ecosystem_category\": \"Plants\",\n",
      "  \"ecosystem_type\": \"NEW ECOSYSTEM_TYPE 2\",\n",
      "  \"ecosystem_subtype\": \"NEW ECOSYSTEM_SUBTYPE 2\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"NEW RAW NAME 2\",\n",
      "    \"name\": \"NEW PI NAME 2\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\",\n",
      "  \"websites\": [\n",
      "    \"HTTP://TEST3.EXAMPLE.COM\",\n",
      "    \"HTTP://TEST4.EXAMPLE.COM\"\n",
      "  ],\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"NEW DOI 2\"\n",
      "  },\n",
      "  \"has_credit_associations\": [\n",
      "    {\n",
      "      \"applied_role\": \"NEW CO-INVESTIGATOR 2\",\n",
      "      \"applies_to_person\": {\n",
      "        \"name\": \"NEW CO-INVESTIGATOR NAME 2\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"applied_role\": \"NEW CURATOR 2\",\n",
      "      \"applies_to_person\": {\n",
      "        \"name\": \"NEW CURATOR NAME 2\"\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n",
      "\n",
      "============== gold:Gs0114663 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"Riverbed sediment microbial communities from the Columbia River, Washington, USA\",\n",
      "  \"description\": \"A metagenomic study to couple microbial communities to carbon and contaminant biogeochemistry in the groundwater-surface water interaction zone (hyporheic zone).\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"Freshwater\",\n",
      "  \"ecosystem_subtype\": \"River\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"James Stegen\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487765\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114663\",\n",
      "  \"name\": \"NEW STUDY NAME 1\",\n",
      "  \"description\": \"NEW DESCRIPTION 1\",\n",
      "  \"ecosystem\": \"NEW ECOSYSTEM 1\",\n",
      "  \"ecosystem_category\": \"Aquatic\",\n",
      "  \"ecosystem_type\": \"NEW ECOSYSTEM_TYPE 1\",\n",
      "  \"ecosystem_subtype\": \"NEW ECOSYSTEM_SUBTYPE 1\",\n",
      "  \"specific_ecosystem\": \"Sediment\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"NEW RAW NAME 1\",\n",
      "    \"name\": \"NEW PI NAME 1\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"NEW DOI 1\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\",\n",
      "  \"websites\": [\n",
      "    \"HTTP://TEST1.EXAMPLE.COM\",\n",
      "    \"HTTP://TEST2.EXAMPLE.COM\"\n",
      "  ],\n",
      "  \"has_credit_associations\": [\n",
      "    {\n",
      "      \"applied_role\": \"NEW CO-INVESTIGATOR 1\",\n",
      "      \"applies_to_person\": {\n",
      "        \"name\": \"NEW CO-INVESTIGATOR NAME 1\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"applied_role\": \"NEW CURATOR 1\",\n",
      "      \"applies_to_person\": {\n",
      "        \"name\": \"NEW CURATOR NAME 1\"\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('test.study_set' was unexpected)\n"
     ]
    }
   ],
   "source": [
    "print_results(process_changesheet(sheetDf, mongodb, temp_db))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1874d3b-cd72-466e-b037-4f96a7352cbe",
   "metadata": {},
   "source": [
    "### set multiple values for credit associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f299543e-9857-4a3a-a23c-72b3d95c134c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>action</th>\n",
       "      <th>attribute</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gold:Gs0114675</td>\n",
       "      <td>update</td>\n",
       "      <td>has_credit_associations</td>\n",
       "      <td>ca1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ca1</td>\n",
       "      <td>update</td>\n",
       "      <td>applied_role</td>\n",
       "      <td>Conceptualization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ca1</td>\n",
       "      <td>update</td>\n",
       "      <td>applies_to_person.name</td>\n",
       "      <td>Kelly Wrighton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ca1</td>\n",
       "      <td>update</td>\n",
       "      <td>applies_to_person.email</td>\n",
       "      <td>Kelly.Wrighton@colostate.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ca1</td>\n",
       "      <td>update</td>\n",
       "      <td>applies_to_person.orcid</td>\n",
       "      <td>orcid:0000-0003-0434-4217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  action                attribute                         value\n",
       "0  gold:Gs0114675  update  has_credit_associations                           ca1\n",
       "1             ca1  update             applied_role             Conceptualization\n",
       "2             ca1  update   applies_to_person.name                Kelly Wrighton\n",
       "3             ca1  update  applies_to_person.email  Kelly.Wrighton@colostate.edu\n",
       "4             ca1  update  applies_to_person.orcid     orcid:0000-0003-0434-4217"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"data/changesheet-array-item-nested-attributes.tsv\", sep=\"\\t\", dtype=\"string\").fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f6141148-801a-4ad4-b90c-3dbe08b0db26",
   "metadata": {},
   "outputs": [],
   "source": [
    "sheetDf = load_changesheet(\"data/changesheet-array-item-nested-attributes.tsv\", mongodb)\n",
    "# sheetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ae62731f-1650-407c-a151-fce526ec2567",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============== gold:Gs0114675 ==============\n",
      "------------------ BEFORE ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114675\",\n",
      "  \"name\": \"Deep subsurface shale carbon reservoir microbial communities from Ohio and West Virginia, USA\",\n",
      "  \"description\": \"Despite the importance of carbon-rich deep shale systems, little is known about microbe-catalyzed processes that may occur naturally within these formations, and how these change as a result of engineered activities. This has important implications on biodiversity change within the deep terrestrial biosphere, the efficiency of energy recovery, and the sustainability of our energy infrastructure. Although limited studies have identified diverse active microorganisms in deep shale formations, more research is needed to understand microbial mechanisms for growth and survival, and how such activities impact carbon and other biogeochemical cycles.\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Terrestrial\",\n",
      "  \"ecosystem_type\": \"Deep subsurface\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Kelly Wrighton\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487763\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\"\n",
      "}\n",
      "------------------ AFTER ------------------\n",
      "{\n",
      "  \"id\": \"gold:Gs0114675\",\n",
      "  \"name\": \"Deep subsurface shale carbon reservoir microbial communities from Ohio and West Virginia, USA\",\n",
      "  \"description\": \"Despite the importance of carbon-rich deep shale systems, little is known about microbe-catalyzed processes that may occur naturally within these formations, and how these change as a result of engineered activities. This has important implications on biodiversity change within the deep terrestrial biosphere, the efficiency of energy recovery, and the sustainability of our energy infrastructure. Although limited studies have identified diverse active microorganisms in deep shale formations, more research is needed to understand microbial mechanisms for growth and survival, and how such activities impact carbon and other biogeochemical cycles.\",\n",
      "  \"ecosystem\": \"Environmental\",\n",
      "  \"ecosystem_category\": \"Terrestrial\",\n",
      "  \"ecosystem_type\": \"Deep subsurface\",\n",
      "  \"ecosystem_subtype\": \"Unclassified\",\n",
      "  \"specific_ecosystem\": \"Unclassified\",\n",
      "  \"principal_investigator\": {\n",
      "    \"has_raw_value\": \"Kelly Wrighton\"\n",
      "  },\n",
      "  \"doi\": {\n",
      "    \"has_raw_value\": \"10.25585/1487763\"\n",
      "  },\n",
      "  \"type\": \"nmdc:Study\",\n",
      "  \"has_credit_associations\": [\n",
      "    {\n",
      "      \"applied_role\": \"Conceptualization\",\n",
      "      \"applies_to_person\": {\n",
      "        \"name\": \"Kelly Wrighton\",\n",
      "        \"email\": \"Kelly.Wrighton@colostate.edu\",\n",
      "        \"orcid\": \"orcid:0000-0003-0434-4217\"\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n",
      "------------------ ERRORS ------------------\n",
      "Additional properties are not allowed ('gold.study_set' was unexpected)\n"
     ]
    }
   ],
   "source": [
    "print_results(process_changesheet(sheetDf, mongodb, temp_db))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32c8706c-7949-4b54-b561-8de01c9e4bb7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
