{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "70cab5b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8d7fdcea",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "from pymongo import MongoClient\n",
    "from pymongo.errors import OperationFailure\n",
    "\n",
    "load_dotenv(\"../../.env.localhost\")\n",
    "\n",
    "admin_client = MongoClient(\n",
    "    host=os.getenv(\"MONGO_HOST\"),\n",
    "    username=os.getenv(\"MONGO_USERNAME\"),\n",
    "    password=os.getenv(\"MONGO_PASSWORD\"),\n",
    "    directConnection=True, # connect to host as a standalone, rather than to entire replicaset\n",
    ")\n",
    "reader_client = MongoClient(\n",
    "    host=os.getenv(\"MONGO_HOST\"),\n",
    "    username=os.getenv(\"MONGO_READONLY_USERNAME\"),\n",
    "    password=os.getenv(\"MONGO_READONLY_PASSWORD\"),\n",
    "    directConnection=True, # connect to host as a standalone, rather than to entire replicaset\n",
    ")\n",
    "\n",
    "mdb_src = reader_client[os.getenv(\"MONGO_DBNAME\")]\n",
    "mdb_tgt = admin_client[os.getenv(\"MONGO_DBNAME\") + \"-scratch\"]\n",
    "\n",
    "\n",
    "try:\n",
    "    mdb_src.study_set.insert_one({\"_id\": \"foobar\"})\n",
    "    print(\"NOT OK\")\n",
    "    mdb_src.study_set.delete_one({\"_id\": \"foobar\"})\n",
    "    print(\"state restored\")\n",
    "    raise Exception(\"can write to `mdb_src`!\")\n",
    "except OperationFailure:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ca3d913-d157-44cd-ad1f-4393d4ab64e4",
   "metadata": {},
   "source": [
    "No output from above? Good!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ceb92334",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"Utils\"\"\"\n",
    "\n",
    "from copy import deepcopy\n",
    "from pprint import pprint\n",
    "\n",
    "import fastjsonschema\n",
    "from toolz import dissoc\n",
    "\n",
    "from nmdc_runtime.util import get_nmdc_jsonschema_dict\n",
    "\n",
    "def without_id_patterns(nmdc_jsonschema):\n",
    "    rv = deepcopy(nmdc_jsonschema)\n",
    "    for cls_, spec in rv[\"$defs\"].items():\n",
    "        if \"properties\" in spec:\n",
    "            if \"id\" in spec[\"properties\"]:\n",
    "                spec[\"properties\"][\"id\"].pop(\"pattern\", None)\n",
    "    return rv\n",
    "\n",
    "def strip_oid(doc):\n",
    "    return dissoc(doc, \"_id\")\n",
    "\n",
    "def get_all(mdb=None, collection_name=None) -> list[dict]:\n",
    "    return {collection_name: [strip_oid(d) for d in mdb_src[collection_name].find()]}\n",
    "\n",
    "def get_some(mdb=None, collection_name=None, limit=5) -> dict:\n",
    "    return {collection_name: [strip_oid(d) for d in mdb_src[collection_name].find(limit=limit)]}\n",
    "\n",
    "def get_one(mdb=None, collection_name=None) -> dict:\n",
    "    return strip_oid(mdb_src[collection_name].find_one())\n",
    "\n",
    "nmdc_jsonschema_validate = fastjsonschema.compile(without_id_patterns(get_nmdc_jsonschema_dict()))\n",
    "\n",
    "def nmdc_schema_collection_names() -> set:\n",
    "    return {\n",
    "        k for k, v in get_nmdc_jsonschema_dict()[\"$defs\"][\"Database\"][\"properties\"].items()\n",
    "        if v.get(\"items\",{}).get(\"$ref\")\n",
    "    }\n",
    "\n",
    "def present_src_collections(mdb) -> list:\n",
    "    return sorted(\n",
    "        n for n in (nmdc_schema_collection_names() & set(mdb_src.list_collection_names()))\n",
    "        if mdb_src[n].estimated_document_count()\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3949bb49-eacd-42da-a588-d8441a010445",
   "metadata": {},
   "source": [
    "Reference for anticipated migration logic/config: https://github.com/microbiomedata/nmdc-schema/blob/v7.4.10/nmdc_schema/migrate_3_2_to_7.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a0fbba94-b5bb-44c6-9ccc-28c2d67b58d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from inspect import signature\n",
    "import re\n",
    "\n",
    "from toolz import assoc, assoc_in, dissoc\n",
    "\n",
    "\n",
    "id_pattern = re.compile(r\"(PO|ENVO):\\d+\")\n",
    "curie_pattern = re.compile(r\"[A-Za-z]\\w+:\\w+\")\n",
    "\n",
    "\n",
    "def ensure_term_id(v):\n",
    "    if not (\"term\" in v or \"has_raw_value\" in v):\n",
    "        raise Exception(f\"needs `term` or `has_raw_value`\")\n",
    "    if not (m := re.search(id_pattern, v[\"has_raw_value\"])):\n",
    "        raise Exception(f'{v[\"has_raw_value\"]} does not match a known ID pattern')\n",
    "    \n",
    "    return assoc(v, \"term\", {\"id\": v[\"has_raw_value\"]})\n",
    "\n",
    "\n",
    "def uppercase_curie_prefixes(v):\n",
    "    if not isinstance(v, list):\n",
    "        raise Exception(\"list expected\")\n",
    "    if not all(\":\" in elt for elt in v):\n",
    "        raise Exception(\"CURIEs expected\")\n",
    "    \n",
    "    rv = []\n",
    "    for elt in v:\n",
    "        prefix, localpart = elt.split(\":\", maxsplit=1)\n",
    "        rv.append(prefix.upper() + \":\" + localpart)\n",
    "    return rv\n",
    "\n",
    "def lowercase_curie_prefixes(v):\n",
    "    if not isinstance(v, list):\n",
    "        raise Exception(\"list expected\")\n",
    "    if not all(\":\" in elt for elt in v):\n",
    "        raise Exception(\"CURIEs expected\")\n",
    "    \n",
    "    rv = []\n",
    "    for elt in v:\n",
    "        prefix, localpart = elt.split(\":\", maxsplit=1)\n",
    "        rv.append(prefix.lower() + \":\" + localpart)\n",
    "    return rv\n",
    "\n",
    "def ensure_curies(v):\n",
    "    \"\"\"\n",
    "    >>> ensure_curies(['HTTPS://identifiers.org/gold:Gs0144557'])\n",
    "    ['gold:Gs0144557']\n",
    "    \"\"\"\n",
    "    if not isinstance(v, list):\n",
    "        raise Exception(\"list expected\")\n",
    "    if not all(\":\" in elt for elt in v):\n",
    "        raise Exception(\"CURIE-ables expected\")\n",
    "\n",
    "    return [re.search(curie_pattern, elt).group(0) for elt in v]\n",
    "\n",
    "def ensure_depth_via_depth2(v, d):\n",
    "    if \"depth\" not in d:\n",
    "        raise Exception(\"no `depth` field\")\n",
    "    \n",
    "    depth = d[\"depth\"]\n",
    "    return {\"depth\": depth, \"depth2\": None}\n",
    "#     depth = d[\"depth\"]\n",
    "#     if v['has_unit'] != depth['has_unit']:\n",
    "#         raise Exception(\"disagreement wrt units\")\n",
    "#     if 'has_numeric_value' not in v:\n",
    "#         return {\"depth\n",
    "    \n",
    "    \n",
    "#     if 'has_maximum_numeric_value' in v:\n",
    "#         if v['has_maximum_numeric_value'] != depth2['has_numeric_value']:\n",
    "#             raise Exception(\"disagreement wrt maximum value\")\n",
    "#     else:\n",
    "#         v = assoc(v, \"has_maximum_numeric_value\", depth2[\"has_numeric_value\"])\n",
    "#         if 'has_minimum_numeric_value' not in v:\n",
    "#             # sets missing min value as given numeric value\n",
    "#             v = assoc(v, \"has_minimum_numeric_value\", v[\"has_numeric_value\"])\n",
    "#         else:\n",
    "#             if v[\"has_numeric_value\"] != v[\"has_minimum_numeric_value\"]:\n",
    "#                 # XXX Is this really a problem?\n",
    "#                 raise Exception(\"min value not the same as numeric value\")\n",
    "\n",
    "\n",
    "def remove_type_field(v):\n",
    "    return dissoc(v, \"type\")\n",
    "\n",
    "\n",
    "def ensure_number_in_meters(v):\n",
    "    if isinstance(v, dict):\n",
    "        assert \"has_unit\" in v, \"no units\"\n",
    "        assert v[\"has_unit\"] in (\"meter\", \"metre\"), \"not meters\"\n",
    "        return v[\"has_numeric_value\"]\n",
    "    assert isinstance(v, int | float), \"not a number\"\n",
    "    return v\n",
    "\n",
    "def ensure_gold_sequencing_project_identifiers(v, d):\n",
    "    if not isinstance(v, list):\n",
    "        raise Exception(\"list expected\")\n",
    "    if \"gold_sequencing_project_identifiers\" not in d:\n",
    "        return v\n",
    "\n",
    "    rv = []\n",
    "    for elt in v:\n",
    "        if elt.lower().startswith(\"gold:gb\"):\n",
    "            if d[\"id\"].lower().startswith(\"gold:gp\"):\n",
    "                rv.append(d[\"id\"])\n",
    "            else:\n",
    "                raise Exception(\"no avaialable substitute\")\n",
    "        else:\n",
    "            rv.append(elt)\n",
    "    return {\"gold_sequencing_project_identifiers\": rv}\n",
    "\n",
    "\n",
    "def rename_num_tRNA(v):\n",
    "    return [\n",
    "        change_fieldname(elt, \"num_tRNA\", \"num_t_rna\")\n",
    "        for elt in v if \"num_tRNA\" in elt\n",
    "    ]\n",
    "\n",
    "\n",
    "fieldname_replacements = {\n",
    "    \"biosample_set\": {\n",
    "        \"INSDC_biosample_identifiers\": \"insdc_biosample_identifiers\",\n",
    "        # # \"part_of\": \"sample_link\",\n",
    "        \"identifier\": \"samp_name\",\n",
    "        \"GOLD_sample_identifiers\": \"gold_biosample_identifiers\",\n",
    "        \"env_broad_scale\": [ensure_term_id, remove_type_field],\n",
    "        \"env_local_scale\": [ensure_term_id, remove_type_field],\n",
    "        \"env_medium\": [ensure_term_id, remove_type_field],\n",
    "        # depends on `dict`s being ordered in python 3.8+\n",
    "        \"gold_biosample_identifiers\": [uppercase_curie_prefixes],\n",
    "        \"depth2\": [ensure_depth_via_depth2],\n",
    "        \"lat_lon\": [remove_type_field],\n",
    "        \"elev\": [ensure_number_in_meters],\n",
    "        \"insdc_biosample_identifiers\": [lowercase_curie_prefixes],\n",
    "    },\n",
    "    \"study_set\": {\n",
    "        \"GOLD_study_identifiers\": \"gold_study_identifiers\",\n",
    "        \"gold_study_identifiers\": [ensure_curies, uppercase_curie_prefixes],\n",
    "    },\n",
    "    \"omics_processing_set\": {\n",
    "        \"GOLD_sequencing_project_identifiers\": \"gold_sequencing_project_identifiers\",\n",
    "        \"gold_sequencing_project_identifiers\": [\n",
    "            ensure_gold_sequencing_project_identifiers,\n",
    "            uppercase_curie_prefixes,\n",
    "        ],\n",
    "    },\n",
    "    \"mags_activity_set\": {\n",
    "        \"mags_list\": [rename_num_tRNA],\n",
    "        \"lowDepth_contig_num\": \"low_depth_contig_num\",\n",
    "    },\n",
    "    \"metagenome_assembly_set\": {\n",
    "        \"ctg_L50\": \"ctg_l50\",\n",
    "        \"ctg_L90\": \"ctg_l90\",\n",
    "        \"ctg_N50\": \"ctg_n50\",\n",
    "        \"ctg_N90\": \"ctg_n90\",\n",
    "        \"scaf_L50\": \"scaf_l50\",\n",
    "        \"scaf_L90\": \"scaf_l90\",\n",
    "        \"scaf_N50\": \"scaf_n50\",\n",
    "        \"scaf_N90\": \"scaf_n90\",\n",
    "        \"scaf_l_gt50K\": \"scaf_l_gt50k\",\n",
    "        \"scaf_n_gt50K\": \"scaf_n_gt50k\",\n",
    "        \"scaf_pct_gt50K\": \"scaf_pct_gt50k\",\n",
    "    },\n",
    "    \"read_qc_analysis_activity_set\": {\n",
    "        \"output_read_bases\": \"output_base_count\",\n",
    "        \"input_read_bases\": \"input_base_count\",\n",
    "    },\n",
    "}\n",
    "\n",
    "biosample_id_routing = {\n",
    "    \"emsl\": \"emsl_biosample_identifiers\",\n",
    "    \"gold\": \"gold_biosample_identifiers\",\n",
    "    \"igsn\": \"igsn_biosample_identifiers\",\n",
    "    \"img.taxon\": \"img_identifiers\"\n",
    "}\n",
    "\n",
    "\n",
    "def change_fieldname(d, fieldname, replacement):\n",
    "    return dissoc(assoc(d, replacement, d[fieldname]), fieldname)\n",
    "\n",
    "\n",
    "def replace_fields(d, context):\n",
    "    assert \"collection_name\" in context\n",
    "    for fieldname, replacement in fieldname_replacements[context[\"collection_name\"]].items():\n",
    "        if fieldname in d:\n",
    "            if isinstance(replacement, list):\n",
    "                assert all(callable(r) for r in replacement), \"replacement-list must be all functions\"\n",
    "                for rfun in replacement:\n",
    "                    n_params = len(signature(rfun).parameters)\n",
    "                    if n_params == 1:\n",
    "                        d = assoc(d, fieldname, rfun(d[fieldname]))\n",
    "                    elif n_params == 2:\n",
    "                        mapping_updates = rfun(d[fieldname], d)\n",
    "                        for k, v in mapping_updates.items():\n",
    "                            d = dissoc(d, k) if v is None else assoc(d, k, v)\n",
    "                    else:\n",
    "                        raise Exception(\"replaced fn takes too many parameters\")\n",
    "            else:\n",
    "                d = change_fieldname(d, fieldname, replacement)\n",
    "    return d\n",
    "\n",
    "\n",
    "def iter_validate(mdb=None, collection_name=None, limit=0, xform=None):\n",
    "    print(f\"validating {collection_name}...\")\n",
    "    for d in mdb_src[collection_name].find(limit=limit):\n",
    "        d = strip_oid(d)\n",
    "        if xform:\n",
    "            d = xform(d, {\"collection_name\": collection_name})\n",
    "        try:\n",
    "            _ = nmdc_jsonschema_validate({collection_name: [d]})\n",
    "        except fastjsonschema.JsonSchemaException as e:\n",
    "            print(d[\"id\"])\n",
    "            print(e)\n",
    "            #pprint(d)\n",
    "            #raise e\n",
    "            return"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64c4a6bf-04c2-42aa-9019-86b127d3aab0",
   "metadata": {},
   "source": [
    "# replacements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f8b8690f-e4b8-4e46-b0d6-fe57125d5f4b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "validating biosample_set...\n",
      "validating data_object_set...\n",
      "validating mags_activity_set...\n",
      "validating metabolomics_analysis_activity_set...\n",
      "nmdc:8969f454c3944f1eac9da499fb950a18\n",
      "data.metabolomics_analysis_activity_set[{data__metabolomicsanalysisactivityset_x}] must not contain {'has_calibration', 'has_metabolite_quantifications'} properties\n",
      "validating metagenome_annotation_activity_set...\n",
      "validating metagenome_assembly_set...\n",
      "validating metagenome_sequencing_activity_set...\n",
      "nmdc:107ade35423143e39dc30b12832ac759\n",
      "data.metagenome_sequencing_activity_set[{data__metagenomesequencingactivityset_x}].git_url must be string\n",
      "validating metaproteomics_analysis_activity_set...\n",
      "nmdc:ec0db6c4103faf47d3393ec27d541639\n",
      "data.metaproteomics_analysis_activity_set[{data__metaproteomicsanalysisactivityset_x}] must not contain {'has_peptide_quantifications'} properties\n",
      "validating metatranscriptome_activity_set...\n",
      "validating nom_analysis_activity_set...\n",
      "nmdc:a1453a2ae5bf4bcca2165b90b96d1082\n",
      "data.nom_analysis_activity_set[{data__nomanalysisactivityset_x}] must not contain {'has_calibration'} properties\n",
      "validating omics_processing_set...\n",
      "validating read_qc_analysis_activity_set...\n",
      "validating study_set...\n"
     ]
    }
   ],
   "source": [
    "for coll_name in present_src_collections(mdb_src):\n",
    "    if coll_name in fieldname_replacements:\n",
    "        iter_validate(mdb_src, coll_name, xform=replace_fields)\n",
    "    else:\n",
    "        iter_validate(mdb_src, coll_name, xform=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d8ef1c8-22f1-47d3-ba2f-5a35939f9367",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (nmdc-runtime)",
   "language": "python",
   "name": "nmdc-runtime"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
